index.js

```diff
@@ -8,59 +8,85 @@ function getStream(inputStream, opts) {
 	}
 
 	var stream;
-	var encoding = (opts && opts.encoding) || 'utf8';
-	var buffer = encoding === 'buffer';
-	var array = encoding === 'array';
+	var array = opts && opts.array;
+	var encoding = opts && opts.encoding;
+	var maxBuffer = opts && opts.maxBuffer;
 
-	if (buffer || array) {
-		encoding = null;
+	if (typeof maxBuffer !== 'number') {
+		maxBuffer = Infinity;
 	}
 
-	var len = 0;
-	var ret = encoding ? '' : [];
+	var buffer = encoding === 'buffer';
+	var objectMode = false;
 
-	function onData(chunk) {
-		if (buffer) {
-			len += chunk.length;
-		}
+	if (array) {
+		objectMode = !(encoding || buffer);
+	} else {
+		encoding = encoding || 'utf8';
+	}
 
-		if (encoding) {
-			ret += chunk;
-		} else {
-			ret.push(chunk);
-		}
+	if (buffer) {
+		encoding = null;
 	}
 
+	var len = 0;
+	var ret = [];
+	var clean;
+
 	var p = new Promise(function (resolve, reject) {
-		stream = new PassThrough({objectMode: array});
+		stream = new PassThrough({objectMode: objectMode});
 		inputStream.pipe(stream);
 
 		if (encoding) {
 			stream.setEncoding(encoding);
 		}
 
+		var onData = function (chunk) {
+			ret.push(chunk);
+
+			if (objectMode) {
+				len = ret.length;
+			} else {
+				len += chunk.length;
+			}
+
+			if (len > maxBuffer) {
+				reject(new Error('maxBuffer exceeded'));
+			}
+		};
+
 		stream.on('data', onData);
 		stream.on('error', reject);
 		stream.on('end', resolve);
-	});
 
-	var clean = function () {
-		stream.removeListener('data', onData);
-	};
+		clean = function () {
+			stream.removeListener('data', onData);
+		};
+	});
 
 	p.then(clean, clean);
 
 	return p.then(function () {
-		return buffer ? Buffer.concat(ret, len) : ret;
+		if (array) {
+			return ret;
+		}
+		return buffer ? Buffer.concat(ret, len) : ret.join('');
 	});
 }
 
 module.exports = getStream;
 
-module.exports.buffer = function (stream) {
-	return getStream(stream, {encoding: 'buffer'});
+module.exports.buffer = function (stream, opts) {
+	return getStream(stream, {
+		encoding: 'buffer',
+		maxBuffer: opts && opts.maxBuffer
+	});
```

Comment:
Should I protect against this?:

```js
getStream.buffer(stream, {encoding:'utf8'});
```

The above doesn't make a lot of sense, but it would mean getting back a string instead of a buffer